core/src/main/scala/com/metamx/tranquility/tranquilizer/Tranquilizer.scala

```diff
@@ -0,0 +1,310 @@
+    */
+  @LifecycleStop
+  def stop(): Unit = {
+    started = false
+    sendThread.interrupt()
+    sendThread.join()
+    Await.result(beam.close())
+  }
+
+  override def toString = s"Tranquilizer($beam)"
+
+  private def requireStarted() {
+    if (!started) {
+      throw new IllegalStateException("Not started")
+    }
+  }
+
+  // Swap out the current buffer, returning it and incrementing pendingBatches. Will block if more than
+  // maxPendingBatches are already pending.
+  private def swap(): Buffer[MessageHolder] = {
+    lock.synchronized {
+      while (pendingBatches >= maxPendingBatches) {
+        if (log.isDebugEnabled) {
+          log.debug(s"pendingBatches[$pendingBatches] >= maxPendingBatches[$maxPendingBatches], waiting...")
+        }
+        lock.wait()
+      }
+
+      val _buffer = buffer
+      buffer = new ArrayBuffer[MessageHolder]()
+      bufferStartMillis = 0L
+      pendingBatches += 1
+      lock.notifyAll()
+      if (log.isDebugEnabled) {
+        log.debug(s"Swapping out buffer with ${_buffer.size} messages, $pendingBatches batches pending.")
+      }
+      _buffer
+    }
+  }
+
+  // Send a buffer of messages. Decrement pendingBatches once the send finishes.
+  private def sendBuffer(myBuffer: Buffer[MessageHolder]): Unit = {
+    if (log.isDebugEnabled) {
+      log.debug(s"Sending buffer with ${myBuffer.size} messages.")
+    }
+
+    val propagate = try {
+      beam.propagate(myBuffer.map(_.message))
+    }
+    catch {
+      case e: Exception =>
+        // Should not happen- exceptions should be in the Future. This is a defensive check.
+        Future.exception(new IllegalStateException("Propagate call failed", e))
+    }
+
+    propagate respond { result =>
+      result match {
+        case Return(n) if n == myBuffer.size =>
+          log.debug(s"Sent $n out of ${myBuffer.size} messages. Marking batch complete.")
+          myBuffer.foreach(_.future.setValue(()))
+
+        case Return(n) =>
+          log.debug(s"Sent $n out of ${myBuffer.size} messages. Marking batch dropped.")
+          val e = new MessageDroppedException(s"Batch not complete ($n/${myBuffer.size})")
+          myBuffer.foreach(_.future.setException(e))
+
+        case Throw(e) =>
+          log.warn(e, s"Failed to send ${myBuffer.size} messages.")
+          myBuffer.foreach(_.future.setException(e))
+      }
+
+      lock.synchronized {
+        pendingBatches -= 1
+        lock.notifyAll()
+        if (log.isDebugEnabled) {
+          log.debug(s"Sent buffer, $pendingBatches batches pending.")
+        }
+      }
+    }
+  }
+}
+
+/**
+  * Exception indicating that a message was dropped "on purpose" by the beam. This is not a recoverable exception
+  * and so the message must be discarded.
+  */
+class MessageDroppedException(message: String) extends Exception(message)
+
+object Tranquilizer
+{
+  val DefaultMaxBatchSize      = 2000
+  val DefaultMaxPendingBatches = 5
```

Comment:
just curious how did you came to these defaults for batchSize and pendingBatches? 
are they based on any properties on underlying beam ?